File: katas/merge_sort.py
def merge(arr1, arr2):
    results = []
    index1 = 0
    index2 = 0

    while index1 < len(arr1) and index2 < len(arr2):
        if arr1[index1] <= arr2[index2]:
            results.append(arr1[index1])
            index1 += 1
        else:
            results.append(arr2[index2])
            index2 += 1

    results.extend(arr1[index1:])
    results.extend(arr2[index2:])

    return results


def mergeSort(array):
    if len(array) <= 1:
        return array

    mid = round(len(array) / 2)
    left_array = mergeSort(array[:mid])
    right_array = mergeSort(array[mid:])

    return merge(left_array, right_array)


print(mergeSort([1,2,8,7,-12,14,120,0,-5,35]))

File: sorting/merge_sort.js
const merge = (arr1, arr2) => {
  const results = [];
  let index1 = 0;
  let index2 = 0;

  while(index1 < arr1.length && index2 < arr2.length) {
    if (arr1[index1] <= arr2[index2]) {
      results.push(arr1[index1])
      index1 ++
    }
    else {
      results.push(arr2[index2])
      index2 ++
    }
  }

/*   while(index1 < arr1.length) {
    results.push(arr1[index1])
    index1++
  }

  while(index2 < arr2.length) {
    results.push(arr2[index2])
    index2++
  } */

  results.push(...arr1.slice(index1), ...arr2.slice(index2));
  return results;
}


const mergeSort = (arr) => {
  if(arr.length <= 1) return arr;
  let midpoint = Math.floor(arr.length/2);
  const leftArr = mergeSort(arr.slice(0,mid));
  const rightArr = mergeSort(arr.slice(mid));
  return merge(leftArr, rightArr)
}